state/protocol/invalid/epoch.go

```diff
@@ -0,0 +1,72 @@
+package invalid
+
+import (
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/state/protocol"
+)
+
+// Epoch represents an epoch that does not exist or could not be retrieved.
+type Epoch struct {
+	err error
+}
+
+func (u *Epoch) Counter() (uint64, error) {
+	return 0, u.err
+}
+
+func (u *Epoch) FirstView() (uint64, error) {
+	return 0, u.err
+}
+
+func (u *Epoch) FinalView() (uint64, error) {
+	return 0, u.err
+}
+
+func (u *Epoch) InitialIdentities() (flow.IdentityList, error) {
+	return nil, u.err
+}
+
+func (u *Epoch) Clustering() (flow.ClusterList, error) {
+	return nil, u.err
+}
+
+func (u *Epoch) Cluster(uint) (protocol.Cluster, error) {
+	return nil, u.err
+}
+
+func (u *Epoch) DKG() (protocol.DKG, error) {
+	return nil, u.err
+}
+
+func (u *Epoch) RandomSource() ([]byte, error) {
+	return nil, u.err
+}
+
+func (u *Epoch) Seed(...uint32) ([]byte, error) {
+	return nil, u.err
+}
+
+func NewEpoch(err error) *Epoch {
+	return &Epoch{err: err}
+}
+
+// Epochs is an epoch query for an invalid snapshot.
+type Epochs struct {
+	err error
+}
+
+func (u *Snapshot) Epochs() protocol.EpochQuery {
+	return &Epochs{err: u.err}
+}
+
+func (u *Epochs) Current() protocol.Epoch {
+	return NewEpoch(u.err)
+}
+
+func (u *Epochs) Next() protocol.Epoch {
+	return NewEpoch(u.err)
```

Comment:
can this recursively go through? (e.g., `..Next().Next().Next()`), if so, does this make sense to prefix the error each time we create a new epoch?

Reply:
No there's only one layer of querying. From a block reference (ie `state.AtBlockID(id)`) we can query only the previous/current/next epochs from that reference.